File: python/linkedlist/leetcode/add_two_numbers.py

# 两数相加

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode()
        cur, p, q, count = dummy, l1, l2, 0
        while p or q:
            if p: p, count = p.next, count + p.val
            if q: q, count = q.next, count + q.val
            cur.next = ListNode(val=count % 10)
            cur, count = cur.next, count // 10
        if count > 0:
            cur.next = ListNode(count)
        return dummy.next

File: go/linkedlist/leetcode/add_two_numbers.go
package main

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{Val: 0}
	cur, p, q, count := dummy, l1, l2, 0
	for p != nil || q != nil {
		if p != nil {
			p, count = p.Next, count+p.Val
		}
		if q != nil {
			q, count = q.Next, count+q.Val
		}
		cur.Next = &ListNode{Val: count % 10}
		cur, count = cur.Next, count/10
	}
	if count != 0 {
		cur.Next = &ListNode{Val: count}
	}
	return dummy.Next
}
